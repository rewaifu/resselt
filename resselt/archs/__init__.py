import importlib.util
import os

from resselt.registry.registry import Registry, Architecture

global_registry = Registry()
base_dir = os.path.dirname(__file__)
base_class = Architecture

for root, dirs, files in os.walk(base_dir):
    if root == base_dir:
        continue

    for file in files:
        if file.endswith('.py'):
            rel_dir = os.path.relpath(root, base_dir)
            module_name = os.path.join(rel_dir, file[:-3]).replace(os.path.sep, '.')
            full_module_name = f'{__name__}.{module_name}'
            module = importlib.import_module(full_module_name)

            for attribute_name in dir(module):
                attribute = getattr(module, attribute_name)
                if isinstance(attribute, type) and issubclass(attribute, base_class) and attribute is not base_class:
                    globals()[attribute_name] = attribute

                    instance = attribute()
                    global_registry.add(instance)
