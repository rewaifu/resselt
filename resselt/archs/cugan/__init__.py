from typing import Mapping, Union, Literal

from .arch import UpCunet2x, UpCunet2x_fast, UpCunet3x, UpCunet4x
import torch
from resselt.registry.key_condition import KeyCondition
from resselt.registry.architecture import WrappedModel, Architecture

_CUGAN = Union[UpCunet2x, UpCunet3x, UpCunet4x, UpCunet2x_fast]


class CUGANArch(Architecture[_CUGAN]):
    def __init__(self):
        super().__init__(
            id='CuGAN',
            detect=KeyCondition.has_all(
                'unet1.conv1.conv.0.weight',
                'unet1.conv1.conv.2.weight',
                'unet1.conv1_down.weight',
                'unet1.conv2.conv.0.weight',
                'unet1.conv2.conv.2.weight',
                'unet1.conv2.seblock.conv1.weight',
                'unet1.conv2_up.weight',
                'unet1.conv_bottom.weight',
                'unet2.conv1.conv.0.weight',
                'unet2.conv1_down.weight',
                'unet2.conv2.conv.0.weight',
                'unet2.conv2.seblock.conv1.weight',
                'unet2.conv3.conv.0.weight',
                'unet2.conv3.seblock.conv1.weight',
                'unet2.conv3_up.weight',
                'unet2.conv4.conv.0.weight',
                'unet2.conv4_up.weight',
                'unet2.conv5.weight',
                'unet2.conv_bottom.weight',
            ),
        )

    def load(self, state_dict: Mapping[str, object]) -> WrappedModel:
        scale: Literal[2, 3, 4]
        in_channels: int
        out_channels: int
        pro: bool = False

        tags: list[str] = []
        if 'pro' in state_dict:
            pro = True
            tags.append('pro')
            state_dict['pro'] = torch.zeros(1)

        in_channels = state_dict['unet1.conv1.conv.0.weight'].shape[1]

        if 'conv_final.weight' in state_dict and in_channels == 12:
            # UpCunet2x_fast
            scale = 2
            in_channels = 3  # hard coded in UpCunet2x_fast
            out_channels = 3  # hard coded in UpCunet2x_fast
            model = UpCunet2x_fast(in_channels=in_channels, out_channels=out_channels)
            tags.append('fast')
        elif 'conv_final.weight' in state_dict:
            # UpCunet4x
            scale = 4
            out_channels = 3  # hard coded in UpCunet4x
            model = UpCunet4x(in_channels=in_channels, out_channels=out_channels, pro=pro)
        elif state_dict['unet1.conv_bottom.weight'].shape[2] == 5:
            # UpCunet3x
            scale = 3
            out_channels = state_dict['unet2.conv_bottom.weight'].shape[0]
            model = UpCunet3x(in_channels=in_channels, out_channels=out_channels, pro=pro)
        else:
            # UpCunet2x
            scale = 2
            out_channels = state_dict['unet2.conv_bottom.weight'].shape[0]
            model = UpCunet2x(in_channels=in_channels, out_channels=out_channels, pro=pro)

        return WrappedModel(model=model, in_channels=in_channels, out_channels=out_channels, upscale=scale, name='CUGAN')
