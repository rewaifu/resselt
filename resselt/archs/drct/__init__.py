import math
from typing import Mapping

from .arch import DRCT
from ...factory import Architecture, KeyCondition
from ...utilities.state_dict import get_seq_len, get_pixelshuffle_params


class MoSRArch(Architecture[DRCT]):
    def __init__(self):
        super().__init__(
            uid='DRCT',
            detect=KeyCondition.has_all(
                'conv_first.weight',
                'conv_first.bias',
                'layers.0.swin1.norm1.weight',
                'layers.0.swin1.norm1.bias',
                'layers.0.swin1.attn.relative_position_bias_table',
                'layers.0.swin1.attn.relative_position_index',
                'layers.0.swin1.attn.qkv.weight',
                'layers.0.swin1.attn.proj.weight',
                'layers.0.swin1.attn.proj.bias',
                'layers.0.swin1.norm2.weight',
                'layers.0.swin1.mlp.fc1.weight',
                'layers.0.swin1.mlp.fc1.bias',
                'layers.0.swin1.mlp.fc2.weight',
                'layers.0.adjust1.weight',
                'layers.0.swin2.norm1.weight',
                'layers.0.adjust2.weight',
                'layers.0.swin3.norm1.weight',
                'layers.0.adjust3.weight',
                'layers.0.swin4.norm1.weight',
                'layers.0.adjust4.weight',
                'layers.0.swin5.norm1.weight',
                'layers.0.adjust5.weight',
                'norm.weight',
                'norm.bias',
            ),
        )

    def load(self, state_dict: Mapping[str, object]):
        patch_size = 1  # cannot be detected
        img_range = 1.0  # cannot be deduced from state_dict

        # detect
        in_chans = state_dict['conv_first.weight'].shape[1]
        embed_dim = state_dict['conv_first.weight'].shape[0]

        num_layers = get_seq_len(state_dict, 'layers')
        depths = (6,) * num_layers
        num_heads = []
        for i in range(num_layers):
            num_heads.append(state_dict[f'layers.{i}.swin1.attn.relative_position_bias_table'].shape[1])

        mlp_ratio = state_dict['layers.0.swin1.mlp.fc1.weight'].shape[0] / embed_dim

        window_square = state_dict['layers.0.swin1.attn.relative_position_bias_table'].shape[0]
        window_size = (math.isqrt(window_square) + 1) // 2

        if 'conv_last.weight' in state_dict:
            upsampler = 'pixelshuffle'
            upscale, _ = get_pixelshuffle_params(state_dict, 'upsample')
        else:
            upsampler = ''
            upscale = 1

        if 'conv_after_body.weight' in state_dict:
            resi_connection = '1conv'
        else:
            resi_connection = 'identity'

        qkv_bias = 'layers.0.swin1.attn.qkv.bias' in state_dict
        gc = state_dict['layers.0.adjust1.weight'].shape[0]

        patch_norm = 'patch_embed.norm.weight' in state_dict
        ape = 'absolute_pos_embed' in state_dict

        if 'layers.0.swin2.attn_mask' in state_dict:
            img_size = math.isqrt(state_dict['layers.0.swin2.attn_mask'].shape[0]) * window_size * patch_size
        else:
            # we only know that the input size is <= window_size,
            # so we just assume that the input size is window_size
            img_size = window_size * patch_size

        model = DRCT(
            img_size=img_size,
            patch_size=patch_size,
            in_chans=in_chans,
            embed_dim=embed_dim,
            depths=depths,
            num_heads=num_heads,
            window_size=window_size,
            mlp_ratio=mlp_ratio,
            qkv_bias=qkv_bias,
            ape=ape,
            patch_norm=patch_norm,
            upscale=upscale,
            img_range=img_range,
            upsampler=upsampler,
            resi_connection=resi_connection,
            gc=gc,
        )

        return self._enhance_model(model=model, in_channels=in_chans, out_channels=in_chans, upscale=upscale, name='DRCT')
