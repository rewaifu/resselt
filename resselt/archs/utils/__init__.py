from .dysample import DySample
from .drop_path import DropPath
from .torch_internals import to_2tuple
from .padding import pad_to_multiple
from . import block

__all__ = ['DySample', 'DropPath', 'to_2tuple', 'pad_to_multiple', 'block']
