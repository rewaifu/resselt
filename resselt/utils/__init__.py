from .tiler import MaxTiler, ExactTiler, NoTiling
from .state_dict import canonicalize_state_dict, get_seq_len, pixelshuffle_scale, dysample_scale, get_pixelshuffle_params
from .tensor import image2tensor, tensor2image, empty_cuda_cache, UpscaleDType
from .upscaler import upscale_with_tiler

__all__ = [
    'MaxTiler',
    'ExactTiler',
    'NoTiling',
    'canonicalize_state_dict',
    'get_seq_len',
    'image2tensor',
    'tensor2image',
    'empty_cuda_cache',
    'upscale_with_tiler',
    'pixelshuffle_scale',
    'dysample_scale',
    'get_pixelshuffle_params',
    'UpscaleDType',
]
