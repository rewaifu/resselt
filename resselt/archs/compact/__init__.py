from typing import Mapping

from .arch import SRVGGNetCompact
from resselt.utils import get_seq_len, pixelshuffle_scale
from resselt.registry.key_condition import KeyCondition
from resselt.registry.architecture import WrappedModel, Architecture


class CompactArch(Architecture[SRVGGNetCompact]):
    def __init__(self):
        super().__init__(
            id='Compact',
            detect=KeyCondition.has_all(
                'body.0.weight',
                'body.1.weight',
            ),
        )

    def load(self, state_dict: Mapping[str, object]) -> WrappedModel:
        state = state_dict

        highest_num = get_seq_len(state, 'body') - 1

        in_nc = state['body.0.weight'].shape[1]
        num_feat = state['body.0.weight'].shape[0]
        num_conv = (highest_num - 2) // 2

        pixelshuffle_shape = state[f'body.{highest_num}.bias'].shape[0]

        scale = pixelshuffle_scale(pixelshuffle_shape, in_nc)
        model = SRVGGNetCompact(
            num_in_ch=in_nc,
            num_out_ch=in_nc,
            num_feat=num_feat,
            num_conv=num_conv,
            upscale=scale,
        )

        return WrappedModel(model=model, in_channels=in_nc, out_channels=in_nc, upscale=scale, name='Compact')
