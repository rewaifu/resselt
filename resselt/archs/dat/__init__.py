import math
from typing import Mapping

from .arch import DAT
from resselt.utils import get_seq_len, pixelshuffle_scale
from resselt.registry.key_condition import KeyCondition
from resselt.registry.architecture import WrappedModel, Architecture


class DatArch(Architecture[DAT]):
    def __init__(self):
        super().__init__(
            id='dat',
            detect=KeyCondition.has_all(
                'conv_first.weight',
                'before_RG.1.weight',
                'before_RG.1.bias',
                'layers.0.blocks.0.norm1.weight',
                'layers.0.blocks.0.norm2.weight',
                'layers.0.blocks.0.ffn.fc1.weight',
                'layers.0.blocks.0.ffn.sg.norm.weight',
                'layers.0.blocks.0.ffn.sg.conv.weight',
                'layers.0.blocks.0.ffn.fc2.weight',
                'layers.0.blocks.0.attn.qkv.weight',
                'layers.0.blocks.0.attn.proj.weight',
                'layers.0.blocks.0.attn.dwconv.0.weight',
                'layers.0.blocks.0.attn.dwconv.1.running_mean',
                'layers.0.blocks.0.attn.channel_interaction.1.weight',
                'layers.0.blocks.0.attn.channel_interaction.2.running_mean',
                'layers.0.blocks.0.attn.channel_interaction.4.weight',
                'layers.0.blocks.0.attn.spatial_interaction.0.weight',
                'layers.0.blocks.0.attn.spatial_interaction.1.running_mean',
                'layers.0.blocks.0.attn.spatial_interaction.3.weight',
                'layers.0.blocks.0.attn.attns.0.rpe_biases',
                'layers.0.blocks.0.attn.attns.0.relative_position_index',
                'layers.0.blocks.0.attn.attns.0.pos.pos_proj.weight',
                'layers.0.blocks.0.attn.attns.0.pos.pos1.0.weight',
                'layers.0.blocks.0.attn.attns.0.pos.pos3.0.weight',
                'norm.weight',
            ),
        )

    def load(self, state_dict: Mapping[str, object]) -> WrappedModel:
        img_size = 64  # cannot be deduced from state dict in general
        split_size = [2, 4]
        upscale = 2

        in_chans = state_dict['conv_first.weight'].shape[1]
        embed_dim = state_dict['conv_first.weight'].shape[0]

        # num_layers = len(depth)
        num_layers = get_seq_len(state_dict, 'layers')
        depth = [get_seq_len(state_dict, f'layers.{i}.blocks') for i in range(num_layers)]

        # num_heads is linked to depth
        num_heads = [2] * num_layers
        for i in range(num_layers):
            if depth[i] >= 2:
                # that's the easy path, we can directly read the head count
                num_heads[i] = state_dict[f'layers.{i}.blocks.1.attn.temperature'].shape[0]
            else:
                # because of a head_num // 2, we can only reconstruct even head counts
                key = f'layers.{i}.blocks.0.attn.attns.0.pos.pos3.2.weight'
                num_heads[i] = state_dict[key].shape[0] * 2

        upsampler = 'pixelshuffle' if 'conv_last.weight' in state_dict else 'pixelshuffledirect'
        resi_connection = '1conv' if 'conv_after_body.weight' in state_dict else '3conv'

        if upsampler == 'pixelshuffle':
            upscale = 1
            for i in range(0, get_seq_len(state_dict, 'upsample'), 2):
                num_feat = state_dict[f'upsample.{i}.weight'].shape[1]
                shape = state_dict[f'upsample.{i}.weight'].shape[0]
                upscale *= int(math.sqrt(shape // num_feat))
        elif upsampler == 'pixelshuffledirect':
            num_feat = state_dict['upsample.0.weight'].shape[1]
            upscale = pixelshuffle_scale(state_dict['upsample.0.weight'].shape[0], in_chans)

        qkv_bias = 'layers.0.blocks.0.attn.qkv.bias' in state_dict

        expansion_factor = float(state_dict['layers.0.blocks.0.ffn.fc1.weight'].shape[0] / embed_dim)

        if 'layers.0.blocks.2.attn.attn_mask_0' in state_dict:
            attn_mask_0_x, attn_mask_0_y, _attn_mask_0_z = state_dict['layers.0.blocks.2.attn.attn_mask_0'].shape

            img_size = int(math.sqrt(attn_mask_0_x * attn_mask_0_y))

        if 'layers.0.blocks.0.attn.attns.0.rpe_biases' in state_dict:
            split_sizes = state_dict['layers.0.blocks.0.attn.attns.0.rpe_biases'][-1] + 1
            split_size = [int(x) for x in split_sizes]

        model = DAT(
            img_size=img_size,
            in_chans=in_chans,
            embed_dim=embed_dim,
            split_size=split_size,
            depth=depth,
            num_heads=num_heads,
            expansion_factor=expansion_factor,
            qkv_bias=qkv_bias,
            upscale=upscale,
            resi_connection=resi_connection,
            upsampler=upsampler,
        )

        return WrappedModel(model=model, in_channels=in_chans, out_channels=in_chans, upscale=upscale, name='DAT')
