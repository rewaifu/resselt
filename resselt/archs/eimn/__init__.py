import re
from typing import Mapping

from .arch import eimn
from resselt.utils import get_seq_len, pixelshuffle_scale
from resselt.registry.key_condition import KeyCondition
from resselt.registry.architecture import WrappedModel, Architecture


class eimnArch(Architecture[eimn]):
    def __init__(self):
        super().__init__(
            id='eimn',
            detect=KeyCondition.has_all(
                'head.0.weight',
                'head.0.bias',
                'tail.0.weight',
                'tail.0.bias',
                'block1.0.layer_scale_1',
                'block1.0.layer_scale_2',
                'block1.0.norm1.weight',
                'block1.0.norm1.bias',
                'block1.0.norm1.running_mean',
                'block1.0.norm1.running_var',
                'block1.0.norm1.num_batches_tracked',
                'block1.0.attn.region.weight',
                'block1.0.attn.region.bias',
                'block1.0.attn.spatial_1.weight',
                'block1.0.attn.spatial_1.bias',
                'block1.0.attn.spatial_2.weight',
                'block1.0.attn.spatial_2.bias',
                'block1.0.attn.fusion.weight',
                'block1.0.attn.fusion.bias',
                'block1.0.attn.proj_value.0.weight',
                'block1.0.attn.proj_value.0.bias',
                'block1.0.attn.proj_query.0.weight',
                'block1.0.attn.proj_query.0.bias',
                'block1.0.attn.out.weight',
                'block1.0.attn.out.bias',
                'block1.0.norm2.weight',
                'block1.0.norm2.bias',
                'block1.0.norm2.running_mean',
                'block1.0.norm2.running_var',
                'block1.0.norm2.num_batches_tracked',
                'block1.0.mlp.linear_in.weight',
                'block1.0.mlp.linear_in.bias',
                'block1.0.mlp.SAL.weight',
                'block1.0.mlp.SAL.bias',
                'block1.0.mlp.linear_out.weight',
                'block1.0.mlp.linear_out.bias',
                'block1.0.mlp.DFFM.norm.weight',
                'block1.0.mlp.DFFM.norm.bias',
                'block1.0.mlp.DFFM.global_reduce.weight',
                'block1.0.mlp.DFFM.global_reduce.bias',
                'block1.0.mlp.DFFM.local_reduce.weight',
                'block1.0.mlp.DFFM.local_reduce.bias',
                'block1.0.mlp.DFFM.channel_expand.weight',
                'block1.0.mlp.DFFM.channel_expand.bias',
                'block1.0.mlp.DFFM.spatial_expand.weight',
                'block1.0.mlp.DFFM.spatial_expand.bias',
                'norm1.weight',
                'norm1.bias',
            ),
        )

    def load(self, state: Mapping[str, object]) -> WrappedModel:
        pattern = r'block(\d+)'
        numbers = [int(re.search(pattern, s).group(1)) for s in state.keys() if re.search(pattern, s)]
        num_stages = max(numbers)
        depths = get_seq_len(state, 'block1')
        mlp_ratio_shape = state['block1.0.mlp.linear_in.weight'].shape
        mlp_ratio = mlp_ratio_shape[0] // 2 / mlp_ratio_shape[1]
        embed_dim = state['head.0.weight'].shape[0]
        scale = pixelshuffle_scale(state['tail.0.weight'].shape[0], 3)

        model = eimn(
            embed_dims=embed_dim,
            scale=scale,
            depths=depths,
            mlp_ratios=mlp_ratio,
            num_stages=num_stages,
        )

        return WrappedModel(model=model, in_channels=3, out_channels=3, upscale=scale, name='EIMN')
