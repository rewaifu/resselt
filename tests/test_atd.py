from resselt.archs.atd import ATD, ATDArch
from tests.utils import ModelAsset, assert_loads_correctly, assert_image_inference, disallowed_props, ImageAssets


def test_atd_load():
    assert_loads_correctly(
        ATDArch(),
        lambda: ATD(),
        lambda: ATD(in_chans=4, embed_dim=60),
        lambda: ATD(window_size=4),
        lambda: ATD(depths=(4, 6, 8, 7, 5), num_heads=(4, 6, 8, 12, 5)),
        lambda: ATD(num_tokens=32, reducted_dim=3, convffn_kernel_size=7, mlp_ratio=3),
        lambda: ATD(qkv_bias=False),
        lambda: ATD(patch_norm=False),
        lambda: ATD(ape=True),
        lambda: ATD(resi_connection='1conv'),
        lambda: ATD(resi_connection='3conv'),
        lambda: ATD(upsampler='', upscale=1),
        lambda: ATD(upsampler='nearest+conv', upscale=4),
        lambda: ATD(upsampler='pixelshuffle', upscale=1),
        lambda: ATD(upsampler='pixelshuffle', upscale=2),
        lambda: ATD(upsampler='pixelshuffle', upscale=3),
        lambda: ATD(upsampler='pixelshuffle', upscale=4),
        lambda: ATD(upsampler='pixelshuffle', upscale=8),
        lambda: ATD(upsampler='pixelshuffledirect', upscale=1),
        lambda: ATD(upsampler='pixelshuffledirect', upscale=2),
        lambda: ATD(upsampler='pixelshuffledirect', upscale=3),
        lambda: ATD(upsampler='pixelshuffledirect', upscale=4),
        lambda: ATD(upsampler='pixelshuffledirect', upscale=8),
    )


def test_atd_inference(snapshot):
    asset = ModelAsset(filename='4x_DWTP_DS_ATDl.pth', url='https://public.yor.ovh/pytorch_models/4x_DWTP_DS_ATDl.pth')
    wrapped_model = asset.load_wrapped_model(ATDArch())

    assert wrapped_model == snapshot(exclude=disallowed_props)
    assert isinstance(wrapped_model.model, ATD)

    assert_image_inference(wrapped_model, asset.filename, ImageAssets.COLOR_CAT_120_113)
