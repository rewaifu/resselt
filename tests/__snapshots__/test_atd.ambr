# serializer version: 1
# name: test_atd_inference
  WrappedModel(
    in_channels=3,
    name='ATD',
    out_channels=3,
    upscale=4,
  )
# ---
